### DISCUSSION TITLE
refetch and react hooks

### DISCUSSION THREAD

**Discussion by 596121109** [TanStack/query#3664] [Q&A]
const [data, setData] = useData({
    pageSize: 10,
    pageNo: 1,
    cn: "",
    total: 0,
    list: [],
    version: 0,
    visible: false,
    formVersion: 0,
  });

 const queryList = useQuery(["lists", { pageNo, pageSize, cn }], () => { console.log(cn); getData(pageNo, pageSize, cn) }, {
    // refetchOnWindowFocus: false,
    // retry: true,
    // keepPreviousData: true,
    enabled: false
  })

  const handleSearch = (cns) => {
    // clearTimeout(timer)
    setData({ cn: cns, version: Date.now(), pageNo: 1 });
    refetch()
    // timer = setTimeout(() => refetch(), 40)
  }

When "enabled" is false, the method refetch is called. The CN field always reads the last value. I guess it is caused by the react hook closure trap. How to solve it

**Comment by 596121109** [comment]
const { data: ndata = { data: { list: [], total: 0 } }, isFetching, refetch } = queryList;

**Comment by TkDodo** [comment]
don't call `refetch`. just call `setData`, and set `enabled` to something like `!!data` to automatically enable the query when data is available.
